mlflow/models/evaluation/base.py

```diff
@@ -456,12 +514,143 @@ def _start_run_or_reuse_active_run(run_id):
+                    "If `evaluators` argument is an evaluator name list, evaluator_config "
+                    "must be a dict contains mapping from evaluator name to individual "
+                    "evaluator config dict."
+                )
+        # Use `OrderedDict.fromkeys` to deduplicate elements but keep elements order.
+        evaluator_name_list = list(OrderedDict.fromkeys(evaluators))
+        evaluator_name_to_conf_map = evaluator_config or {}
+    else:
+        raise ValueError(
+            "`evaluators` argument must be None, an evaluator name string, or a list of "
+            "evaluator names."
+        )
+
+    return evaluator_name_list, evaluator_name_to_conf_map
+
+
+_last_failed_evaluator = None
+
+
+def _get_last_failed_evaluator():
+    """
+    Return the evaluator name of the last failed evaluator when calling `evalaute`.
+    This can be used to check which evaluator fail when `evaluate` API fail.
+    """
+    return _last_failed_evaluator
+
+
+def _evaluate(
+    *, model, model_type, dataset, actual_run_id, evaluator_name_list, evaluator_name_to_conf_map
+):
+    """
+    The public API "evaluate" will verify argument first, and then pass normalized arguments
+    to the _evaluate method.
+    """
+    # import _model_evaluation_registry and PyFuncModel inside function to avoid circuit importing
+    from mlflow.models.evaluation.evaluator_registry import _model_evaluation_registry
+
+    global _last_failed_evaluator
+    _last_failed_evaluator = None
+
+    client = mlflow.tracking.MlflowClient()
+    model_uuid = model.metadata.model_uuid
+    dataset._log_dataset_tag(client, actual_run_id, model_uuid)
+
+    eval_results = []
+    for evaluator_name in evaluator_name_list:
+        config = evaluator_name_to_conf_map.get(evaluator_name) or {}
+        try:
+            evaluator = _model_evaluation_registry.get_evaluator(evaluator_name)
+        except MlflowException:
+            _logger.warning(f"Evaluator '{evaluator_name}' is not registered.")
+            continue
+
+        _last_failed_evaluator = evaluator_name
+        if evaluator.can_evaluate(model_type, config):
+            _logger.info(f"Evaluating the model with the {evaluator_name} evaluator.")
+            result = evaluator.evaluate(model, model_type, dataset, actual_run_id, config)
+            eval_results.append(result)
+
+    _last_failed_evaluator = None
+
+    if len(eval_results) == 0:
+        raise ValueError(
+            "The model could not be evaluated by any of the registered evaluators, please "
+            "verify that the model type and other configs are set correctly."
+        )
+
+    merged_eval_result = EvaluationResult(EvaluationMetrics(), dict())
+    for eval_result in eval_results:
+        merged_eval_result.metrics.update(eval_result.metrics)
+        merged_eval_result.artifacts.update(eval_result.artifacts)
+
+    return merged_eval_result
+
+
+@experimental
 def evaluate(
+    *,
```

Comment:
What does the "*" mean here? Do you intend to make it a nameless positional arg like *args? It may worth to spell it out even if it's not used, and put it in the end of the argument list, e.g.
def _evaluate(model, model_type, dataset, actual_run_id, evaluator_name_list, evaluator_name_to_conf_map, *unused_args):

Reply:
No. The `*` is a python grammar, not a argument, it means all following arguments must be a key-word-only arguments. https://www.python.org/dev/peps/pep-3102/

Mark all argument as  key-word-only is a proposal raised by @dbczumar , so that in future if mlflow containerized model provide "model_type" property, we can remove the "model_type" argument here but keep backwards compatibility

Reply:
Wait, removing the `model_type` argumnet is a breaking change, right?

Reply:
In future, we can change API to be :
```
def evaluate(
    *, model, dataset, run_id, evaluators=None, evaluator_config=None, **kwargs
):
   if "model_type" in kwargs:
      model_type = kwargs["kwargs"]
      logger.warning("deprecation: ....")
   else:
      model_type = model.model_type
   ...
```
Then we can keep backwards compatiblity.

Reply:
I didn't know about this usage pattern. Interesting!
Another way to keep backward compatibility is to use keyword arg for model_type, e.g.
```
def evaluate(
    model, dataset, run_id, model_type=None, evaluators=None, evaluator_config=None, **kwargs
):
```
What's the downside of going this more "conventional" way of specifying optional arguments?

Reply:
In current `evaluate` implementation, the `model_type` argument is required. In future we will make it "optional and deprecated" and then remove this argument.